### ISSUE TITLE
Enhancement: [no-unnecessary-condition] brainstorm better usage with constants

### ISSUE THREAD

**Issue by Zamiell** [typescript-eslint/typescript-eslint#5634]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-condition/

### The Problem

The `no-unnecessary-condition` rule is a fantastic rule that has caught many bugs in my codebase. However, it unfortunately causes a lot of false positives when used with static constants.

Consider the following pattern:

```ts
const IS_DEV = true;

// Later on in the code...
if (IS_DEV) {
  // Do something.
}
```

Or, perhaps something like this, which is similar:

```ts
const LOGGING_LEVEL= "warn";

// Later on in the code...
if (LOGGING_LEVEL === "warn") {
  // Do something.
}
```

Both of these examples will cause false positives for the `no-unnecessary-condition` rule, for obvious reasons. This pattern causes me to litter my codebase with eslint-disable comments.

### Why Not Use Environment Variables?

Your gut response to this problem might be something along the lines of:
"This style of code is an anti-pattern. Constants like this should be expressed as environment variables, and not hard-coded into the source code itself."

Unfortunately, environment variables can't be used everywhere, like the browser.

### Annotate Constants

It strikes me that instead of writing N eslint-disable comments for each individual constant, it would be more declarative to annotate the constant itself with a parsable JSDoc comment. The advantage of this is that we only have to do 1 annotation (instead of N annotations).

For example:

```ts
/** @constant */
const IS_DEV = true;

// Later on in the code...
if (IS_DEV) {
  // Do something.
}
```

Then, the rule would automatically ignore any conditional statement with the constant.

I am not particularly biased towards any particular solution; I more wanted to put forward the proposition that it should be less painful to use constants with the `no-unnecessary-condition` rule in general.

**Comment by bradzacher** [comment]
You can hoist the globals into a separate file and make them truly global like
```ts
declare global {
  export const LOGGING_LEVEL: string;
}

global.LOGGING_LEVEL = 'warn';
```

In your app entrypoint you import this file like `import './constants'`. Then anywhere in your code you can reference `LOGGING_LEVEL`.

---

You can actually use environment variables in both node and web. Most bundler support the concept of global substitution - meaning they let you replace a string of code with another.

I.e. You can make the bundler replace the code string `process.env.LOGGING_LEVEL` with the code `'warn'`.

This technique is great because the replacement happens before minification - meaning tools like terser can optimise away always false code and save you bytes from your bundle!

---

The final option is to annotate the variables with assertions.

```ts
const LOGGING_LEVEL = 'warn' as string;
```

Sadly you can't use a variable annotation because TS does some magic optimisation to ignore it on `const` variables :(

This assertion forces TS to use the wider type, which would then prevent our rule from reporting on it.

**Comment by Zamiell** [comment]
Both of those solutions seem pretty good to me. Thank you Brad for the thorough reply.
